includes/class-database-cache.php

```diff
@@ -83,6 +83,49 @@ class Database_Cache implements MultiCurrencyCacheInterface {
 	 */
 	private $refresh_disabled;
 
+	/**
+	 * The maximum size in bytes of a cache entry that will be stored in memory.
+	 */
+	const IN_MEMORY_CACHE_ENTRY_SIZE_LIMIT = 15 * KB_IN_BYTES;
+
+	/**
+	 * The maximum data size in bytes of all the in-memory cache stored data.
+	 *
+	 * This will ensure we keep the memory impact under control.
+	 * This is especially important for large stores that can handle many concurrent requests.
+	 */
+	const IN_MEMORY_CACHE_SIZE_LIMIT = 100 * KB_IN_BYTES;
+
+	/**
+	 * In-memory cache for the duration of a single request.
+	 *
+	 * This is used to avoid multiple database reads for the same data and as a backstop in case the database write fails,
+	 * thus ensuring the cache generator (which means an API call to our platform) is not called multiple times.
+	 *
+	 * @var array
+	 */
+	private $in_memory_cache = [];
+
+	/**
+	 * Cache sizes in bytes for each entry in the in-memory cache.
+	 *
+	 * This is used to keep track of the size of the in-memory cache and stop adding entries if the cache size limit is reached.
+	 * The cache size is calculated based on the serialized size of the data (the same format as the database cache).
+	 *
+	 * We don't use an eviction policy for the in-memory cache because there is no priority among different entries.
+	 * Under normal circumstances, the cache size limit should be sufficient for all entries small enough to be stored in memory.
+	 *
+	 * @var array
+	 */
+	private $in_memory_cache_sizes = [];
```

Comment:
Is this an optimization to avoid repeatedly calculating the overall cache size using `strlen`?

Reply:
If we end up complying with @marcinbot's suggestion of removing per-entry size restriction, we could switch to an integer variable:

```suggestion
	private int $in_memory_cache_current_size;
```

Reply:
> Is this an optimization to avoid repeatedly calculating the overall cache size using strlen?

Yes. That was the intent. Also, to be able to delete entries from cache and remove their size from the total with clarity.

Reply:
> removing per-entry size restriction, we could switch to an integer variable:

Marcin proposed that we remove all cache size management.